java.nio.channels.ClosedChannelException

On Ubuntu 18.04, running ETL with the usual command line:

mvn exec:java -Dexec.mainClass="org.heigit.bigspatialdata.oshdb.tool.importer.extract.Extract" -Dexec.args="--pbf /home/ertert/oshdb-new/oshdb-tool/etl/files/switzerland-internal.osh.pbf -tmpDir ./tmpFiles -workDir ./results --timevalidity_from 2017-01-01" -Dexec.cleanupDaemonThreads=false

when execution starts I get:

joining on thread Thread[org.heigit.bigspatialdata.oshdb.tool.importer.extract.Extract.main(),5,org.heigit.bigspatialdata.oshdb.tool.importer.extract.Extract]
extracting key tables ...[WARNING] 
java.lang.RuntimeException: java.nio.channels.ClosedChannelException

which becomes :

BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.887 s
[INFO] Finished at: 2020-11-26T08:24:50Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:java (default-cli) on project etl: An exception occured while executing the Java class. java.nio.channels.ClosedChannelException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:java (default-cli) on project etl: An exception occured while executing the Java class. java.nio.channels.ClosedChannelException

java.nio.channels.ClosedChannelException is a fairly common issue and generally related to programming. Any thoughts?

--- a/oshdb-tool/etl/src/main/java/org/heigit/bigspatialdata/oshdb/tool/importer/load/LoaderKeyTables.java
+++ b/oshdb-tool/etl/src/main/java/org/heigit/bigspatialdata/oshdb/tool/importer/load/LoaderKeyTables.java
@@ -60,11 +60,10 @@ public class LoaderKeyTables {
         values = new ArrayList<>(kvp.valuesNumber);
 
         valuesChannel.position(kvp.valuesOffset);
-        try(DataInputStream valueStream = new DataInputStream(Channels.newInputStream(valuesChannel));){
-          for (int j = 0; j < kvp.valuesNumber; j++) {
-            final VF vf = VF.read(valueStream);
-            values.add(vf.value);
-          }
+        DataInputStream valueStream = new DataInputStream(Channels.newInputStream(valuesChannel));
+        for (int j = 0; j < kvp.valuesNumber; j++) {
+          final VF vf = VF.read(valueStream);
+          values.add(vf.value);
         }
 
         handler.loadKeyValues(i, key, values);

--- a/oshdb-tool/etl/src/main/java/org/heigit/bigspatialdata/oshdb/tool/importer/transform/TransformerTagRoles.java
+++ b/oshdb-tool/etl/src/main/java/org/heigit/bigspatialdata/oshdb/tool/importer/transform/TransformerTagRoles.java
@@ -10,7 +10,6 @@ import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.OutputStream;
 import java.io.RandomAccessFile;
 import java.nio.channels.Channels;
 import java.nio.channels.FileChannel;
@@ -104,29 +103,27 @@ public class TransformerTagRoles {
     final Int2IntAVLTreeMap valueHash2Cnt = new Int2IntAVLTreeMap();
 
     channel.position(kvp.valuesOffset);
-    try(DataInputStream valueStream = new DataInputStream(new BufferedInputStream(Channels.newInputStream(channel),1024*1024));){
-      for (int j = 0; j < kvp.valuesNumber; j++) {
-        final VF vf = VF.read(valueStream);
-        final int hash = hashFunction.applyAsInt(vf.value);
-        valueHash2Cnt.addTo(hash, 1);
-      }
+    DataInputStream valueStream = new DataInputStream(new BufferedInputStream(Channels.newInputStream(channel),1024*1024));
+    for (int j = 0; j < kvp.valuesNumber; j++) {
+      final VF vf = VF.read(valueStream);
+      final int hash = hashFunction.applyAsInt(vf.value);
+      valueHash2Cnt.addTo(hash, 1);
     }
 
     final Int2IntMap uniqueValues = new Int2IntAVLTreeMap();
     final Object2IntMap<String> notUniqueValues = new Object2IntAVLTreeMap<String>();
     long estimatedSize = 0;
     channel.position(kvp.valuesOffset);
-    try(DataInputStream valueStream = new DataInputStream(Channels.newInputStream(channel));){
-      for (int j = 0; j < kvp.valuesNumber; j++) {
-        final VF vf = VF.read(valueStream);
-        final int hash = hashFunction.applyAsInt(vf.value);
-        if (valueHash2Cnt.get(hash) > 1) {
-          notUniqueValues.put(vf.value, j);
-          estimatedSize += SizeEstimator.estimatedSizeOfAVLEntryValue(kvp.key)+4;
-        } else {
-          uniqueValues.put(hash, j);
-          estimatedSize += SizeEstimator.avlTreeEntry() + 8;
-        }
+    valueStream = new DataInputStream(Channels.newInputStream(channel));
+    for (int j = 0; j < kvp.valuesNumber; j++) {
+      final VF vf = VF.read(valueStream);
+      final int hash = hashFunction.applyAsInt(vf.value);
+      if (valueHash2Cnt.get(hash) > 1) {
+        notUniqueValues.put(vf.value, j);
+        estimatedSize += SizeEstimator.estimatedSizeOfAVLEntryValue(kvp.key)+4;
+      } else {
+        uniqueValues.put(hash, j);
+        estimatedSize += SizeEstimator.avlTreeEntry() + 8;
       }
     }
 

--- a/oshdb-tool/oshpbf-parser/src/main/java/org/heigit/bigspatialdata/oshpbf/parser/rx/RxOshPbfReader.java
+++ b/oshdb-tool/oshpbf-parser/src/main/java/org/heigit/bigspatialdata/oshpbf/parser/rx/RxOshPbfReader.java
@@ -187,7 +187,6 @@ public class RxOshPbfReader {
         }
         pushBackBytes[3] = (byte) nextByte;
       }
-    
       nextByte = pushBackStream.read();
       totalBytesRead++;
     }

--- a/oshdb-tool/oshpbf-parser/src/main/java/org/heigit/bigspatialdata/oshpbf/parser/rx/RxOshPbfReader.java
+++ b/oshdb-tool/oshpbf-parser/src/main/java/org/heigit/bigspatialdata/oshpbf/parser/rx/RxOshPbfReader.java
@@ -190,7 +190,7 @@ public class RxOshPbfReader {
       nextByte = pushBackStream.read();
       totalBytesRead++;
     }
-// we reach the end of stream and found no signature
+    // we reach the end of stream and found no signature
     return -1;
   }
 

--- a/oshdb-tool/oshpbf-parser/src/main/java/org/heigit/bigspatialdata/oshpbf/parser/rx/RxOshPbfReader.java
+++ b/oshdb-tool/oshpbf-parser/src/main/java/org/heigit/bigspatialdata/oshpbf/parser/rx/RxOshPbfReader.java
@@ -142,59 +142,57 @@ public class RxOshPbfReader {
   public static long seekBlobHeaderStart(final InputStream is, final long limit) throws IOException {
     long totalBytesRead = 0;
     final byte[] pushBackBytes = new byte[BLOBHEADER_SIZE_BYTES + SIGNATURE_SIZE_BYTES];
-    try(PushbackInputStream pushBackStream = new PushbackInputStream(is, pushBackBytes.length);) {
-      for (int i = 0; i < BLOBHEADER_SIZE_BYTES; i++) {
-        pushBackBytes[i] = (byte) pushBackStream.read();
-        totalBytesRead++;
-      }
-  
-      int nextByte = pushBackStream.read();
+    PushbackInputStream pushBackStream = new PushbackInputStream(is, pushBackBytes.length);
+    for (int i = 0; i < BLOBHEADER_SIZE_BYTES; i++) {
+      pushBackBytes[i] = (byte) pushBackStream.read();
       totalBytesRead++;
-      int val = 0;
-      while (nextByte != -1) {
-        if ((val < SIGNATURE_OSMDATA.length && SIGNATURE_OSMDATA[val] == nextByte)
-            || (val < SIGNATURE_OSMHEADER.length && SIGNATURE_OSMHEADER[val] == nextByte)) {
-  
+    }
+
+    int nextByte = pushBackStream.read();
+    totalBytesRead++;
+    int val = 0;
+    while (nextByte != -1) {
+      if ((val < SIGNATURE_OSMDATA.length && SIGNATURE_OSMDATA[val] == nextByte)
+          || (val < SIGNATURE_OSMHEADER.length && SIGNATURE_OSMHEADER[val] == nextByte)) {
+    
+        pushBackBytes[BLOBHEADER_SIZE_BYTES + val] = (byte) nextByte;
+    
+        if (val == SIGNATURE_OSMDATA.length - 1 || val == SIGNATURE_OSMHEADER.length - 1) {
+          // Full OSMHeader\Data SIGNATURE is found.
+          pushBackStream.unread(pushBackBytes, 0, BLOBHEADER_SIZE_BYTES + val + 1);
+          totalBytesRead -= BLOBHEADER_SIZE_BYTES + val + 1;
+          return totalBytesRead;
+        }
+        val++;
+      } else if (val != 0) {
+        // break
+        if (limit > 0 && totalBytesRead > limit) {
+          return -1;
+        }
+    
+        val = 0;
+        if (SIGNATURE_OSMDATA[val] == nextByte || SIGNATURE_OSMHEADER[val] == nextByte) {
           pushBackBytes[BLOBHEADER_SIZE_BYTES + val] = (byte) nextByte;
-  
-          if (val == SIGNATURE_OSMDATA.length - 1 || val == SIGNATURE_OSMHEADER.length - 1) {
-            // Full OSMHeader\Data SIGNATURE is found.
-            pushBackStream.unread(pushBackBytes, 0, BLOBHEADER_SIZE_BYTES + val + 1);
-            totalBytesRead -= BLOBHEADER_SIZE_BYTES + val + 1;
-            return totalBytesRead;
-          }
           val++;
-        } else if (val != 0) {
-          // break
-          if (limit > 0 && totalBytesRead > limit) {
-            return -1;
-          }
-  
-          val = 0;
-          if (SIGNATURE_OSMDATA[val] == nextByte || SIGNATURE_OSMHEADER[val] == nextByte) {
-            pushBackBytes[BLOBHEADER_SIZE_BYTES + val] = (byte) nextByte;
-            val++;
-          } else {
-            for (int i = 0; i < 3; i++) {
-              pushBackBytes[i] = pushBackBytes[i + 1];
-            }
-            pushBackBytes[3] = (byte) nextByte;
-          }
-  
         } else {
           for (int i = 0; i < 3; i++) {
             pushBackBytes[i] = pushBackBytes[i + 1];
           }
           pushBackBytes[3] = (byte) nextByte;
         }
-  
-        nextByte = pushBackStream.read();
-        totalBytesRead++;
+    
+      } else {
+        for (int i = 0; i < 3; i++) {
+          pushBackBytes[i] = pushBackBytes[i + 1];
+        }
+        pushBackBytes[3] = (byte) nextByte;
       }
-  
-      // we reach the end of stream and found no signature
-      return -1;
+    
+      nextByte = pushBackStream.read();
+      totalBytesRead++;
     }
+// we reach the end of stream and found no signature
+    return -1;
   }
 
   private static BiFunction<PbfChannel, Emitter<PbfBlob>, PbfChannel> readBlobFromChannel() {

